File: app/src/main/java/com/keplersegg/myself/Activities/MasterActivity.java
package com.keplersegg.myself.Activities;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Toast;

import com.google.android.gms.auth.api.Auth;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.keplersegg.myself.Async.IHttpProvider;
import com.keplersegg.myself.Helper.TokenType;
import com.keplersegg.myself.Interfaces.IErrorMessage;
import com.keplersegg.myself.Models.User;
import com.keplersegg.myself.MySelfApplication;
import com.keplersegg.myself.R;
import com.keplersegg.myself.Room.AppDatabase;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@SuppressLint("Registered")
public class MasterActivity extends AppCompatActivity
    implements
        GoogleApiClient.OnConnectionFailedListener,
        IHttpProvider,
        IErrorMessage
{

    public MySelfApplication application;
    public MasterActivity master;
    public GoogleApiClient mGoogleApiClient;

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);

        master = this;
        application = (MySelfApplication) getApplication();

        Window window = getWindow();
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
        window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
        window.setStatusBarColor(GetColor(R.color.colorPrimary));

        initializeGPlusSettings();
    }

    public int GetColor(int id) { return ContextCompat.getColor(this, id); }

    @NonNull
    public AppDatabase AppDB() { return AppDatabase.getAppDatabase(this); }

    public void NavigateToActivity(String activityName, boolean clearTop) {

        Intent i = null;

        switch (activityName) {

            case "Main":

                i = new Intent(application, MainActivity.class);
                break;

            case "Login":

                i = new Intent(application, LoginActivity.class);
                break;

        default:
        break;
    }

        if (i != null)
            NavigateToActivity(i, clearTop);
    }

    protected void NavigateToActivity(Intent i, boolean clearTop) {

        if (clearTop) // ana sayfadan geri dön'e tıklandığında çıkış yapma diyaloğu gelmesi için.
            i.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);

        startActivity(i);
        overridePendingTransition(R.anim.fade_in, R.anim.fade_out);

        if (clearTop) // ana sayfadan geri dön'e tıklandığında çıkış yapma diyaloğu gelmesi için.
            finish();
    }

    private void initializeGPlusSettings(){
        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                //.requestIdToken(getString(R.string.google_client_id))
                //.requestServerAuthCode(getString(R.string.google_client_id))
                .requestEmail()
                .build();
        mGoogleApiClient = new GoogleApiClient.Builder(this)
                .enableAutoManage(this, this)
                .addApi(Auth.GOOGLE_SIGN_IN_API, gso)
                .build();
    }

    @Override
    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
        // An unresolvable error has occurred and Google APIs (including Sign-In) will not
        // be available.
    }

    protected boolean handleGPlusSignInResult(GoogleSignInAccount account) {

        setToken(TokenType.Google, account.getIdToken());

        application.user = new User();
        application.user.Email = account.getEmail();
        application.user.FirstName = account.getGivenName();
        application.user.LastName = account.getFamilyName();
        Uri pictureUri = account.getPhotoUrl();
        if (pictureUri != null)
            application.user.PictureUrl = pictureUri.toString();

        return true;
    }

    public void setToken(TokenType tokenType, String token) {

        switch (tokenType) {

            case MySelf:
                application.dataStore.setAccessToken(token);
                break;

            case Facebook:
                application.dataStore.setFacebookToken(token);
                break;

            case Google:
                application.dataStore.setGoogleToken(token);
                break;

            default:
                break;
        }
    }

    @NonNull
    public ConnectivityManager getConnectivityManager() {
        return (ConnectivityManager)getSystemService(Context.CONNECTIVITY_SERVICE);
    }

    @Nullable
    public String getAccessToken() {
        return application.dataStore.getAccessToken();
    }

    @Nullable
    public String getDeviceId() { return application.dataStore.getRegisterID(); }

    @Override
    public void showErrorMessage(@NonNull final String message) {

        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Toast.makeText(MasterActivity.this, message, Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public void logException(@NotNull Exception exception, @NotNull String message) {

        //CrashLogger.AddExceptionLog(message, exc);
        showErrorMessage(message);
    }
}

File: app/src/main/java/com/keplersegg/myself/Activities/MasterActivity.kt
package com.keplersegg.myself.Activities

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.os.Bundle
import android.view.WindowManager
import android.widget.Toast

import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.keplersegg.myself.Interfaces.IHttpProvider
import com.keplersegg.myself.Helper.TokenType
import com.keplersegg.myself.Interfaces.IErrorMessage
import com.keplersegg.myself.Models.User
import com.keplersegg.myself.MySelfApplication
import com.keplersegg.myself.R
import com.keplersegg.myself.Room.AppDatabase
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat

@SuppressLint("Registered")
open class MasterActivity : AppCompatActivity(), IHttpProvider, IErrorMessage {

    var application: MySelfApplication? = null
    var master: MasterActivity? = null
    var mGoogleSigninClient: GoogleSignInClient? = null

    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)

        master = this
        application = getApplication() as MySelfApplication

        val window = window
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
        window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
        window.statusBarColor = GetColor(R.color.colorPrimary)

        initializeGPlusSettings()
    }

    fun GetColor(id: Int): Int {
        return ContextCompat.getColor(this, id)
    }

    fun AppDB(): AppDatabase {
        return AppDatabase.getAppDatabase(this)!!
    }

    fun NavigateToActivity(activityName: String, clearTop: Boolean) {

        var i: Intent? = null

        when (activityName) {

            "Main" ->

                i = Intent(application, MainActivity::class.java)

            "Login" ->

                i = Intent(application, LoginActivity::class.java)

            else -> {
            }
        }

        if (i != null)
            NavigateToActivity(i, clearTop)
    }

    protected fun NavigateToActivity(i: Intent?, clearTop: Boolean) {

        if (clearTop)
        // ana sayfadan geri dön'e tıklandığında çıkış yapma diyaloğu gelmesi için.
            i!!.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)

        startActivity(i)
        overridePendingTransition(R.anim.fade_in, R.anim.fade_out)

        if (clearTop)
        // ana sayfadan geri dön'e tıklandığında çıkış yapma diyaloğu gelmesi için.
            finish()
    }

    private fun initializeGPlusSettings() {
        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                //.requestIdToken(getString(R.string.google_client_id))
                //.requestServerAuthCode(getString(R.string.google_client_id))
                .requestEmail()
                .build()

        mGoogleSigninClient = GoogleSignIn.getClient(this, gso)
    }

    protected fun handleGoogleSignInResult(account: GoogleSignInAccount): Boolean {

        setToken(TokenType.Google, account.idToken)

        application!!.user = User()
        application!!.user!!.Email = account.email
        application!!.user!!.FirstName = account.givenName
        application!!.user!!.LastName = account.familyName
        val pictureUri = account.photoUrl
        if (pictureUri != null)
            application!!.user!!.PictureUrl = pictureUri.toString()

        return true
    }

    open fun setToken(tokenType: TokenType, token: String?) {

        when (tokenType) {

            TokenType.MySelf -> application!!.dataStore!!.setAccessToken(token)

            TokenType.Facebook -> application!!.dataStore!!.setFacebookToken(token)

            TokenType.Google -> application!!.dataStore!!.setGoogleToken(token)

        }
    }

    override fun getConnectivityManager(): ConnectivityManager {
        return getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    }

    override fun getAccessToken(): String? {
        return application!!.dataStore!!.getAccessToken()
    }

    override fun getDeviceId(): String? {
        return application!!.dataStore!!.getRegisterID()
    }

    override fun showErrorMessage(message: String) {

        runOnUiThread { Toast.makeText(this@MasterActivity, message, Toast.LENGTH_SHORT).show() }
    }

    override fun logException(exception: Exception, message: String) {

        //CrashLogger.AddExceptionLog(message, exc);
        showErrorMessage(message)
    }
}
